Assets/HoloToolkit/Input/Scripts/Interactable.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+
+namespace HoloToolkit.Unity
+{
+    /// <summary>
+    /// Base class Interactable GameObjects can inherit from.
+    /// </summary>
+    public class Interactable : MonoBehaviour
+    {
+        protected bool TriggerKeywordOnGazeOnly;
+
+        protected bool IsFocused;
+
+        protected virtual void OnEnable()
+        {
+            GestureManager.Instance.OnTap += OnTap;
+            GazeManager.Instance.OnGazeExit += OnGazeExit;
+            GazeManager.Instance.OnGazeEnter += OnGazeEnter;
+            KeywordManager.Instance.OnKeywordRecognized += KeywordRecognized;
+        }
+
+        protected virtual void OnDisable()
+        {
+            GestureManager.Instance.OnTap -= OnTap;
+            GazeManager.Instance.OnGazeExit -= OnGazeExit;
+            GazeManager.Instance.OnGazeEnter -= OnGazeEnter;
+            KeywordManager.Instance.OnKeywordRecognized -= KeywordRecognized;
+        }
+
+        /// <summary>
+        /// Called when a user has tapped any gameObject.
+        /// </summary>
+        /// <param name="tappedGameObject">GameObject user has tapped.</param>
+        protected virtual void OnTap(GameObject tappedGameObject)
+        {
+            if (tappedGameObject == gameObject)
```

Comment:
I modified the test to include the SendMessage times.  Simply create an empty scene and put this on the camera. delete the directional light. press play.

```
using System;
using System.Diagnostics;
using UnityEngine;
using UnityEngine.Events;

public class TestScript : MonoBehaviour
{
    private const int NumReps = 10000000;
    private const int MaxListeners = 1;
    private const int MaxArgs = 3;

    [Serializable]
    private class IntEvent1 : UnityEvent<int>
    {
    }

    [Serializable]
    private class IntEvent2 : UnityEvent<int, int>
    {
    }

    private event Action csharpEv0;
    private event Action<int> csharpEv1;
    private event Action<int, int> csharpEv2;
    private UnityEvent unityEv0 = new UnityEvent();
    private IntEvent1 unityEv1 = new IntEvent1();
    private IntEvent2 unityEv2 = new IntEvent2();

    private string report;

    private void Start()
    {
        Stopwatch stopwatch = new Stopwatch();
        long[,] csharpEvTimes = new long[MaxArgs, MaxListeners];
        long[,] unityEvTimes = new long[MaxArgs, MaxListeners];
        long[,] sendMsgEvTimes = new long[MaxArgs, MaxListeners];

        for (int numListeners = 0; numListeners < MaxListeners; ++numListeners)
        {
            csharpEv0 += NoOp0;
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                if (csharpEv0 != null)
                {
                    csharpEv0.Invoke();
                }
            }
            csharpEvTimes[0, numListeners] = stopwatch.ElapsedMilliseconds;

            unityEv0.AddListener(NoOp0);
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                unityEv0.Invoke();
            }
            unityEvTimes[0, numListeners] = stopwatch.ElapsedMilliseconds;

            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; j++)
            {
                SendMessage("SendMessage", SendMessageOptions.DontRequireReceiver);
            }
            sendMsgEvTimes[0, numListeners] = stopwatch.ElapsedMilliseconds;
        }
        for (int numListeners = 0; numListeners < MaxListeners; ++numListeners)
        {
            csharpEv1 += NoOp1;
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                if (csharpEv1 != null)
                {
                    csharpEv1.Invoke(11);
                }
            }
            csharpEvTimes[1, numListeners] = stopwatch.ElapsedMilliseconds;

            unityEv1.AddListener(NoOp1);
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                unityEv1.Invoke(11);
            }
            unityEvTimes[1, numListeners] = stopwatch.ElapsedMilliseconds;
        }
        for (int numListeners = 0; numListeners < MaxListeners; ++numListeners)
        {
            csharpEv2 += NoOp2;
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                if (csharpEv2 != null)
                {
                    csharpEv2.Invoke(11, 22);
                }
            }
            csharpEvTimes[2, numListeners] = stopwatch.ElapsedMilliseconds;

            unityEv2.AddListener(NoOp2);
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                unityEv2.Invoke(11, 22);
            }
            unityEvTimes[2, numListeners] = stopwatch.ElapsedMilliseconds;
        }

        report = "Num Args|Num Listeners|C# Event Time (ms)|UnityEvent Time(ms)|SendMessage Time(ms)\n------------ | ------------- | ------------- | ------------- | -------------\n";
        for (int i = 0; i < MaxArgs; ++i)
        {
            for (int j = 0; j < MaxListeners; ++j)
            {
                report += i + "|" + (j + 1) + "|" + csharpEvTimes[i, j] + "|" + unityEvTimes[i, j] + "|" + sendMsgEvTimes[i, j] + "\n";
            }
        }
    }

    private void OnGUI()
    {
        GUI.TextArea(new Rect(0, 0, Screen.width, Screen.height), report);
    }

    private static void NoOp0() { }
    private static void NoOp1(int a) { }
    private static void NoOp2(int a, int b) { }
    private void SendMessage() { }
}
```